### PR TITLE
Feat/557 rest api add oauth2clientcredentials to built in auth methods

### LINKED ISSUE #1870: rest_api: Add `OAuth2ClientCredentials` to the list of built-in auth methods
rest_Add `OAuth2ClientCredentials` to the current list of built-in auth methods.

This requires changes only in the dlt-core repo in rest_client/auth.py

### Tasks:
- [x]  in sources/rest_api/typing.py: add AuthType, add OAuth2Config
- [x] remove `__init()__`
- [x] implement `__post_init__()`
- [x] uncomment skip in tests/rest_api/test_configurations.py::test_auth_type_configs
- [x] update docs

### Specification:
The code should look like as follows:

```py
@configspec
class OAuth2ClientCredentials(OAuth2AuthBase):
    """
    This class implements OAuth2 Client Credentials flow where the autorization service
    gives permission without the end user approving.
    This is often used for machine-to-machine authorization.
    The client sends its client ID and client secret to the authorization service which replies
    with a temporary access token.
    With the access token, the client can access resource services.
    """
    access_token_url: TSecretStrValue = None
    client_id: TSecretStrValue = None
    client_secret: TSecretStrValue = None
    access_token_request_data: Dict[str, Any] = None
    default_token_expiration: int = 3600
    session: Annotated[BaseSession, NotResolved()] = None

    def __post_init__(self) -> None:
        if self.access_token_request_data is None:
            self.access_token_request_data = {}

        self.token_expiry: pendulum.DateTime = pendulum.now()

        if self.session is None:
            self.session = requests.client.session

```

### PR DESCRIPTION
**Pull request by willi-mueller**
### Description
- Adds `oauth2_client_credentials` to auth shorthand in the RESTAPIConfig
- masks the secrets of aouth2 config dict during dict validation

### Related Issues

- Resolves #1870

### CODE REVIEW

File: docs/website/docs/dlt-ecosystem/verified-sources/rest_api/basic.md [hunk 1]
```diff
@@ -512,7 +512,7 @@ Available authentication types:
 | [BearTokenAuth](../../../general-usage/http/rest-client.md#bearer-token-authentication) | `bearer` | Bearer token authentication. |
 | [HTTPBasicAuth](../../../general-usage/http/rest-client.md#http-basic-authentication) | `http_basic` | Basic HTTP authentication. |
 | [APIKeyAuth](../../../general-usage/http/rest-client.md#api-key-authentication) | `api_key` | API key authentication with key defined in the query parameters or in the headers. |
-| [OAuth2ClientCredentials](../../../general-usage/http/rest-client.md#oauth20-authorization) | N/A | OAuth 2.0 authorization with a temporary access token obtained from the authorization server. |
+| [OAuth2ClientCredentials](../../../general-usage/http/rest-client.md#oauth20-authorization) | `oauth2_client_credentials` | OAuth 2.0 authorization with a temporary access token obtained from the authorization server. |
```
Comment (willi-mueller) [inline docs/website/docs/dlt-ecosystem/verified-sources/rest_api/basic.md:515]: Or do we want to be bold and offer `oauth_2`?

Comment (burnash) [reply docs/website/docs/dlt-ecosystem/verified-sources/rest_api/basic.md:515]: No, I think it's good

**Review by burnash — DISMISSED** [review]
Great job, thank you @willi-mueller

**Comment by rudolfix** [comment]
I reimplemented part of this PR in other branch. somehow it is hanging here completed for weeks... once test pass I'll merge it

**Pull request merged**